pkgs/tools/virtualization/nixos-container/nixos-container.pl

```diff
@@ -309,14 +290,14 @@ sub safeRemoveTree {
 }
 
 elsif ($action eq "root-login") {
-    runInContainer("@su@", "root", "-l");
+    exec("machinectl", "shell", $containerName, "/bin/sh", "-l");
 }
 
 elsif ($action eq "run") {
     shift @ARGV; shift @ARGV;
     # Escape command.
     my $s = join(' ', map { s/'/'\\''/g; "'$_'" } @ARGV);
-    runInContainer("@su@", "root", "-l", "-c", "exec " . $s);
+    exec("machinectl", "--quiet", "shell", $containerName, "/bin/sh", "-l", "-c", $s);
```

Comment:
@Mic92 That is exactly how it this works. Notice the line above (`my $s...`), which already escapes the arguments.
```
# nixos-container run vpn echo \$PATH
Connected to machine vpn. Press ^] three times within 1s to exit session.
$PATH
Connection to machine vpn terminated.
```
But this already worked before.